File: graphs/DFS/dfs_stack.py
def dfs(start,graph): # performing DFS using Stack 
    visited=[False]*len(graph) # list to keep a check whetehr the node has been visited or not 
    stack=[] # a stack to perform DFS
    
    stack.append(start) # appending the start point of the DFS
    
    while stack: # interating over the stack
        node= stack.pop() # taking out node from the Stack
        
        if visited[node]: # checking if the node is visited or not 
            continue
        
        visited[node]=True # marking the node as visited
        print(node,end=" ") # printing the node
        
        
        for neighbour in reversed(graph[node]): # iterating over the neighbour of node in the reversed order and appending in the stack 
            if not visited[neighbour]:
                visited[neighbour]=True
                stack.append(neighbour) # appending unvisited neighbour in the stack 
                
n = 5

graph = [[] for _ in range(n)]

graph[0].append(1)
graph[0].append(2)
graph[1].append(3)
graph[2].append(4)

dfs(0, graph)
            

File: graphs/DFS/DFS_stack.java
import java.util.*;

public class DFS_stack {
    public static void main(String[] args) {
        int V = 5;

        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();

        for (int i = 0; i < V; i++) {
            graph.add(new ArrayList<>());
        }

        // Undirected graph
        graph.get(0).add(1);
        graph.get(0).add(2);

        graph.get(1).add(0);
        graph.get(1).add(3);
        graph.get(1).add(4);

        graph.get(2).add(0);

        graph.get(3).add(1);

        graph.get(4).add(1);

        dfs(0, graph);
    }

    static void dfs(int start, ArrayList<ArrayList<Integer>> graph) {
        boolean[] visited = new boolean[graph.size()];
        Stack<Integer> stack = new Stack<>();

        stack.push(start);

        while (!stack.isEmpty()) {
            int node = stack.pop();

            if (!visited[node]) {
                visited[node] = true;
                System.out.print(node + " ");
            }

            for (int i = graph.get(node).size() - 1; i >= 0; i--) {
                int neighbour = graph.get(node).get(i);

                if (!visited[neighbour]) {
                    stack.push(neighbour);
                }
            }
        }
    }
}